pkg/issuer/acme/setup_test.go

```diff
@@ -0,0 +1,520 @@
+	acmecl "github.com/jetstack/cert-manager/pkg/acme/client"
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	cmacme "github.com/jetstack/cert-manager/pkg/apis/acme/v1"
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	v1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	controllertest "github.com/jetstack/cert-manager/pkg/controller/test"
+	"github.com/jetstack/cert-manager/pkg/util"
+	"github.com/jetstack/cert-manager/pkg/util/errors"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+	"github.com/jetstack/cert-manager/test/unit/coreclients"
+	"github.com/jetstack/cert-manager/test/unit/gen"
+)
+
+func TestAcme_Setup(t *testing.T) {
+	var (
+		fixedClockStart = time.Now()
+		fakeclock       = fakeclock.NewFakeClock(fixedClockStart)
+		nowMetaTime     = metav1.NewTime(fakeclock.Now())
+
+		baseIssuer = gen.Issuer("test-issuer",
+			gen.SetIssuerACMEURL(acmev2Prod))
+		// base issuer condition
+		baseCondition = gen.IssuerCondition(v1.IssuerConditionReady,
+			gen.SetIssuerConditionStatus(cmmeta.ConditionFalse),
+			gen.SetIssuerConditionLastTransitionTime(&nowMetaTime))
+		issuerSecretKeyName = "test"
+
+		ecdsaPrivKey = mustGenerateEDCSAKey(t)
+		rsaPrivKey   = mustGenerateRSAKey(t)
+
+		notFoundErr    = apierrors.NewNotFound(corev1.Resource("test"), "test")
+		invalidDataErr = errors.NewInvalidData("test")
+		someErr        = fmt.Errorf("test")
+		invalidURL     = "%"
+		acmeErr450     = &acmeapi.Error{StatusCode: 450}
+		acmeErr500     = &acmeapi.Error{StatusCode: 500}
+		//TODO: we should probably mock calls to net/url instead of doing this.
+		invalidURLErr = parseURLErr(invalidURL)
+
+		invalidURLMessage        = fmt.Sprintf(messageTemplateFailedToParseURL, invalidURL, invalidURLErr)
+		invalidAccountURLMessage = fmt.Sprintf(messageTemplateFailedToParseAccountURL, invalidURL, invalidURLErr)
+
+		someEmail    = "test@test.com"
+		someEmailURL = fmt.Sprintf("mailto:%s", someEmail)
+
+		// to be used where we don't care what value is passed
+		someString = "test"
+
+		// eabSecret is a mock value for secret with EAB key that user would have created.
+		// 'ZEdWemRBbz0K' is 'test' double base64-encoded.
+		// cert-manager only accepts double-encoded values, see https://github.com/jetstack/cert-manager/pull/3877#discussion_r610717791 .
+		eabSecret = gen.Secret(someString,
+			gen.SetSecretData(map[string][]byte{"key": []byte("ZEdWemRBbz0K")}))
+
+		// 'dGVzdAo=\n' is 'ZEdWemRBbz0K' decoded + a newline.
+		// This is the decoded EAB key that we send to the ACME server.
+		// TODO: could the newline cause any issues?
+		eabKey = "dGVzdAo=\n"
+	)
+
+	tests := map[string]struct {
+		issuer v1.GenericIssuer
+
+		// Private key returned by keyFromSecret stub.
+		kfsKey crypto.Signer
+		// Error returned by keyFromSecret stub.
+		kfsErr error
+		// Whether keyFromSecret should be called.
+		kfsShouldBeCalled bool
```

Comment:
Subjective rambling: Personally, I think this is a step too far...by which I mean that I think this is an example of the test reaching too far into the implementation of the function-under-test....does it matter whether the injected `keyFromSecret` is called? Or does it only matter whether what values that function returns ....but I haven't finished reading all the changes yet, so might yet delete this comment.

Reply:
I looked at this again- in this particular case I do agree. It's called from the main path of execution and we do check the returned private key afterwards + it is an implementation detail as such- so there isn't really anything gained from the check. I will remove it.